### ISSUE TITLE
`Neo.DatabaseError.Statement.ExecutionFailure` on combination of ORDER BY and LIMIT 0

### ISSUE THREAD

**Issue by s101d1** [neo4j/neo4j#5545]
I found this bug on Neo4j v2.2.5.

First, insert some sample nodes like:

```
merge (a1:Actor {id: "1", name: "A"})
merge (a2:Actor {id: "2", name: "B"})
merge (a3:Actor {id: "3", name: "C"})
merge (a4:Actor {id: "4", name: "D"})
```

Then run this query:

```
match (a:Actor)
return a
order by a.name
limit 0;
```

I got `Neo.DatabaseError.Statement.ExecutionFailure = 0` error message after running the query.

This bug only occur if there are `ORDER BY` and `LIMIT 0` on the query.

**Comment by thobe** [comment]
I've confirmed that this only happens when combining `ORDER BY` and `LIMIT 0`, it is because the code used for keeping the _top n_ elements is unable to cope with limits smaller than `1`. While this is sensible for normal queries, I can see how a generated query might end up with `LIMIT 0` in it. What we really should do in that case is short-cut the entire query to return an empty result set.

Here is a stack trace (from 2.3) of the root cause:

```
java.lang.ArrayIndexOutOfBoundsException: 0
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.TopPipe.internalCreateResults(TopPipe.scala:83)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:125)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:93)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:63)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:230)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:214)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anon$3.run(ExecutionPlanBuilder.scala:172)
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:192)
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:192)
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.runSafely(CompatibilityFor2_3.scala:114)
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper.run(CompatibilityFor2_3.scala:191)
    at org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:27)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:118)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:112)
```

**Comment by thobe** [comment]
Passing in a negative limit (through a parameter, the parser barfs if trying to use a negative literal) fails in equally bad ways:

```
java.lang.NegativeArraySizeException
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.TopPipe.internalCreateResults(TopPipe.scala:82)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:125)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:93)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:63)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:230)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:214)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anon$3.run(ExecutionPlanBuilder.scala:172)
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:192)
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:192)
    at org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.runSafely(CompatibilityFor2_3.scala:114)
    at org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper.run(CompatibilityFor2_3.scala:191)
    at org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:27)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:118)
    at org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:112)
```

(the query was: `match (n) return n order by n.name limit {lim};`, and `lim` was `-1`)

**Comment by thobe** [comment]
Fix: #5549
